drivers/sub8_videoray_m5_thruster/sub8_thruster_comm/thruster_comm.py

```diff
@@ -1,84 +1,181 @@
-
-        TODO:
-            --> 0x88, 0x8c: Set slew-rate up and down to something tiny
-            --> Send messages without getting status
-            --> Get status without sending thrust
-            --> Determine which thrusters are on a port
         '''
         self.port_name = port_info['port']
         self.port = self.connect_port(self.port_name)
-        # holds the motor_ids of all the thrusters that were supposed to be here and were actually found
-        self.thruster_dict = {}
+        self.CONFIDENT_MODE = port_info['CONFIDENT_MODE']  # Dangerous mode that will not detect thruster losses
+
+        self.thruster_info = {}  # Information about all thrusters declared in the layout
+        self.online_thruster_names = set()  # Keeps track of all thrusters that can be allocated thrust
+
+        # Flag to avoid simultaneous sharing of the serial line b/w thrusters
+        self.serial_busy = False
 
         # Load thruster configurations and check if the requested thruster exists on this port
-        self.missing_thrusters = []
         for thruster_name in port_info['thruster_names']:
-            try:
-                # Note: will only try to detect thrusters as listed in the layout. That means
-                # that if a thruster is connected to the wrong port, IT WILL NOT BE FOUND.
-                self.load_thruster_config(thruster_name, thruster_definitions)
-            except IOError:
-                self.missing_thrusters.append(thruster_name)
-                continue
+            self.load_thruster(thruster_name, thruster_definitions)
+
+        # Setup infrastructure to monitor comms quality
+        num_thrusters = len(self.online_thruster_names)
+        names = self.get_declared_thruster_names()
+        self.command_tx_count = dict.fromkeys(names, 0)  # Commands sent per thruster
+        self.status_rx_count = dict.fromkeys(names, 0)  # Statuses received per thruster
+        self.command_latency_avg = dict.fromkeys(names, rospy.Duration(0))  # Average latency tx rx per thruster
 
     def connect_port(self, port_name):
         '''Connect to and return a serial port'''
         try:
-            serial_port = serial.Serial(port_name, baudrate=self._baud_rate, timeout=self._timeout)
-        except IOError as e:
-            rospy.logerr("Could not connect to thruster port {}".format(port_name))
-            raise(e)
-        return serial_port
-
-    def load_thruster_config(self, thruster_name, thruster_definitions):
-        '''Loads the motor_id from the config file if the thruster was found'''
-        # Get our humungous error string ready
+            serial_port = serial.Serial(port_name, baudrate=self._baud_rate, timeout=self._read_timeout)
+            serial_port.flushInput()
+            return serial_port
+        except serial.serialutil.SerialException as e:
+            rospy.logwarn("Could not connect to thruster port {}".format(port_name))
+            raise SubjuGatorException()
```

Comment:
This would be a good example to add a SubjuGatorException inherited Exception class, like SubjuGatorSerialException